src/components/token/token-controller.ts

```diff
@@ -0,0 +1,126 @@
+  grant_type: "authorization_code";
+  code: string;
+  redirectUri: string;
+  client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
+  client_assertion: string;
+};
+
+export const tokenController = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  try {
+    const config = Config.getInstance();
+    const tokenRequestParameters: TokenRequest = req.body;
+    const tokenRequestValidation = tokenRequestValidator(
+      tokenRequestParameters
+    );
+
+    if (!tokenRequestValidation.valid) {
+      handleTokenRequestError(tokenRequestValidation.error, res);
+      return;
+    }
+
+    const parsedPrivateKeyJwt = privateKeyJwtParser(tokenRequestParameters);
+
+    const privateKeyJwtValidation = await privateKeyJwtValidator(
+      parsedPrivateKeyJwt.privateKeyJwt
+    );
+
+    if (!privateKeyJwtValidation.valid) {
+      handleTokenRequestError(privateKeyJwtValidation.error, res);
+      return;
+    }
+
+    const authCodeParams = config.getAuthCodeRequestParams(
+      tokenRequestParameters.code
+    );
+
+    if (!authCodeParams) {
+      logger.warn(
+        { code: tokenRequestParameters.code },
+        "Could not find auth code params for provided auth code"
+      );
+      throw new TokenRequestError({
+        error_code: "invalid_grant",
+        error_description: "Invalid grant",
+        httpStatusCode: 400,
+      });
+    }
+
+    if (
+      !isRedirectUriValid(
+        authCodeParams.redirectUri,
+        tokenRequestParameters.redirectUri
+      )
+    ) {
+      logger.warn(
+        {
+          authCodeRedirect: authCodeParams.redirectUri,
+          tokenRequestRedirect: tokenRequestParameters.redirectUri,
+        },
+        "Mismatch in redirect uri between auth code params and token request"
+      );
+      throw new TokenRequestError({
+        error_code: "invalid_grant",
+        error_description: "Invalid grant",
+        httpStatusCode: 400,
+      });
+    }
+
+    const accessToken = await createAccessToken(
+      authCodeParams.scopes,
+      authCodeParams.sessionId
+    );
+    const idToken = await createIdToken(authCodeParams, accessToken);
+
+    res.status(200).json({
+      access_token: accessToken,
+      token_type: "Bearer",
+      expires_in: 3600,
+      id_token: idToken,
+    });
+
+    return;
+  } catch (error) {
+    logger.error("Token Request Error: " + (error as Error).message);
+
+    if (error instanceof TokenRequestError) {
+      handleTokenRequestError(error, res);
+      return;
+    } else if (error instanceof ParseRequestError) {
+      res.status(400).json({
+        error: "invalid_request",
+        error_description: "Invalid private_key_jwt",
+      });
+      return;
+    } else {
+      res.status(500).json({
+        error: "server_error",
+      });
+      return;
+    }
+  }
+};
+
+const isRedirectUriValid = (paramStoreUri: string, tokenRequestUri: string) =>
+  paramStoreUri === tokenRequestUri;
```

Comment:
I feel like we can probably just do this inline?

Reply:
Yep can do!

Reply:
Done